dev/PipsPager/PipsPager.cpp

```diff
@@ -0,0 +1,404 @@
+        m_lastMaxDisplayedPages = oldValue;
+        SetVerticalPipsSVMaxSize();
+        UpdateVerticalPips(NumberOfPages(), MaxDisplayedPages());
+        UpdateNavigationButtonVisualStates();
+    }
+    else {
+        MaxDisplayedPages(0);
+    }
+}
+
+void PipsPager::OnNumberOfPagesChanged()
+{
+    SetVerticalPipsSVMaxSize();
+    const int numberOfPages = NumberOfPages();
+    if (SelectedPageIndex() > numberOfPages - 1 && numberOfPages > -1) {
+        SelectedPageIndex(numberOfPages - 1);
+    }
+    else {
+        UpdateVerticalPips(numberOfPages, MaxDisplayedPages());
+        UpdateNavigationButtonVisualStates();
+    }
+}
+
+void PipsPager::OnSelectedPageIndexChanged(const int oldValue)
+{
+    // If we don't have any pages, there is nothing we should do.
+    // Ensure that SelectedPageIndex will end up in the valid range of values
+    // Special case is NumberOfPages being 0, in that case, don't verify upperbound restrictions
+    if (SelectedPageIndex() > NumberOfPages() - 1 && NumberOfPages() > 0)
+    {
+        SelectedPageIndex(NumberOfPages() - 1);
+    }
+    else if (SelectedPageIndex() < 0)
+    {
+        SelectedPageIndex(0);
+    }
+    else {
+        // Now handle the value changes
+        m_lastSelectedPageIndex = oldValue;
+
+        // Fire value property change for UIA
+        if (const auto peer = winrt::FrameworkElementAutomationPeer::FromElement(*this).try_as<winrt::PipsPagerAutomationPeer>())
+        {
+            winrt::get_self<PipsPagerAutomationPeer>(peer)->RaiseSelectionChanged(m_lastSelectedPageIndex, SelectedPageIndex());
+        }
+
+        UpdateVerticalPips(NumberOfPages(), MaxDisplayedPages());
+        UpdateNavigationButtonVisualStates();
+        RaiseSelectedIndexChanged();
+    }
+}
+
+void PipsPager::RaiseSelectedIndexChanged()
+{
+    const auto args = winrt::make_self<PipsPagerSelectedIndexChangedEventArgs>(m_lastSelectedPageIndex, SelectedPageIndex());
+    m_selectedIndexChangedEventSource(*this, *args);
+}
+
+void PipsPager::MovePipIdentifierToElement(int index) {
+
+    if (NumberOfPages() != 0) {
+        if (const auto repeater = m_verticalPipsRepeater.get())
+        {
+            if (const auto element = repeater.TryGetElement(m_lastSelectedPageIndex).try_as<winrt::Button>()) {
+                element.Style(DefaultIndicatorStyle());
+            }
+            if (const auto element = repeater.GetOrCreateElement(index).try_as<winrt::Button>()) {
+                element.Style(SelectedIndicatorStyle());
+                ScrollToCenterOfViewport(element);
+            }
+        }
+    }
+}
+
+void PipsPager::SetVerticalPipsSVMaxSize() {
+    // TODO: Extract the actual height from the first pip in the ItemsRepeater by measuring it
+    // and then getting desiredHeight. While doing it, I encountered a problem that scrollViewer
+    // will not update layout after its MaxHeight or simply Height was changed.
+    // Stephen or Ranjesh - any thoughts?:)
```

Comment:
@ranjeshj @StephenLPeters , thoughts?:)
I also cannot figure out how to leave the 1st element of the repeater always realized so when its size changes, layout can be updated effectively. Any other way to make sure that the size of the scrollViewer is always what it should be?:)